requestmanager/server.go

```diff
@@ -0,0 +1,296 @@
+			DoNotSendCids:        doNotSendCids,
+			NodePrototypeChooser: hooksResult.CustomChooser,
+			ResumeMessages:       resumeMessages,
+			PauseMessages:        pauseMessages,
+		})
+	return request, incoming, incomingError
+}
+
+func (rm *RequestManager) terminateRequest(requestID graphsync.RequestID) {
+	ipr, ok := rm.inProgressRequestStatuses[requestID]
+	if ok {
+		log.Infow("graphsync request complete", "request id", requestID, "peer", ipr.p, "total time", time.Since(ipr.startTime))
+	}
+	delete(rm.inProgressRequestStatuses, requestID)
+	rm.asyncLoader.CleanupRequest(requestID)
+	if ok {
+		for _, onTerminated := range ipr.onTerminated {
+			select {
+			case <-rm.ctx.Done():
+			case onTerminated <- nil:
+			}
+		}
+	}
+}
+
+func (rm *RequestManager) cancelRequest(requestID graphsync.RequestID, isPause bool, onTerminated chan<- error, terminalError error) {
+	inProgressRequestStatus, ok := rm.inProgressRequestStatuses[requestID]
+	if !ok {
+		if onTerminated != nil {
+			select {
+			case onTerminated <- graphsync.RequestNotFoundErr{}:
+			case <-rm.ctx.Done():
+			}
+		}
+		return
+	}
+
+	if onTerminated != nil {
+		inProgressRequestStatus.onTerminated = append(inProgressRequestStatus.onTerminated, onTerminated)
+	}
+	if terminalError != nil {
+		select {
+		case inProgressRequestStatus.terminalError <- terminalError:
+		default:
+		}
+	}
+
+	rm.SendRequest(inProgressRequestStatus.p, gsmsg.CancelRequest(requestID))
+	if isPause {
+		inProgressRequestStatus.paused = true
+	} else {
+		inProgressRequestStatus.cancelFn()
+	}
+}
+
+func (rm *RequestManager) processResponseMessage(p peer.ID, responses []gsmsg.GraphSyncResponse, blks []blocks.Block) {
+	filteredResponses := rm.processExtensions(responses, p)
+	filteredResponses = rm.filterResponsesForPeer(filteredResponses, p)
+	rm.updateLastResponses(filteredResponses)
+	responseMetadata := metadataForResponses(filteredResponses)
+	rm.asyncLoader.ProcessResponse(responseMetadata, blks)
+	rm.processTerminations(filteredResponses)
+}
+
+func (rm *RequestManager) filterResponsesForPeer(responses []gsmsg.GraphSyncResponse, p peer.ID) []gsmsg.GraphSyncResponse {
+	responsesForPeer := make([]gsmsg.GraphSyncResponse, 0, len(responses))
+	for _, response := range responses {
+		requestStatus, ok := rm.inProgressRequestStatuses[response.RequestID()]
+		if !ok || requestStatus.p != p {
+			continue
+		}
+		responsesForPeer = append(responsesForPeer, response)
+	}
+	return responsesForPeer
+}
+
+func (rm *RequestManager) processExtensions(responses []gsmsg.GraphSyncResponse, p peer.ID) []gsmsg.GraphSyncResponse {
+	remainingResponses := make([]gsmsg.GraphSyncResponse, 0, len(responses))
+	for _, response := range responses {
+		success := rm.processExtensionsForResponse(p, response)
+		if success {
+			remainingResponses = append(remainingResponses, response)
+		}
+	}
+	return remainingResponses
+}
+
+func (rm *RequestManager) updateLastResponses(responses []gsmsg.GraphSyncResponse) {
```

Comment:
what this method (amongst others) does seems not fully reflected in the name. some comments on expected behavior of these methods would be useful

Reply:
will add in cleanup PR